tools/tpm2_policysecret.c

```diff
@@ -0,0 +1,191 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <tss2/tss2_sys.h>
+
+#include "files.h"
+#include "log.h"
+#include "tpm2_alg_util.h"
+#include "tpm2_auth_util.h"
+#include "tpm2_options.h"
+#include "tpm2_policy.h"
+#include "tpm2_session.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct tpm2_policysecret_ctx tpm2_policysecret_ctx;
+struct tpm2_policysecret_ctx {
+    //File path for the session context data
+    const char *session_path;
+    //File path for storing the policy digest output
+    const char *out_policy_dgst_path;
+    //Specifying the TPM object handle-id or the file path of context blob
+    const char *context_arg;
+    tpm2_loaded_object context_object;
+    //Auth value of the auth object
+    char *auth_str;
+    struct {
+        UINT8 c : 1;
+        UINT8 p : 1;
+    } flags;
+    TPMS_AUTH_COMMAND session_data;
+};
+
+static tpm2_policysecret_ctx ctx = {
+    .session_data = TPMS_AUTH_COMMAND_INIT(TPM2_RS_PW),
+};
+
+static bool on_option(char key, char *value) {
+
+    bool result = true;
+
+    switch (key) {
+    case 'o':
+        ctx.out_policy_dgst_path = value;
+        break;
+    case 'S':
+        ctx.session_path = value;
+        break;
+    case 'c':
+        ctx.context_arg = value;
+        ctx.flags.c = 1;
+        break;
+    case 'p':
+        ctx.auth_str = value;
+        ctx.flags.p = 1;
+        break;
+    }
+
+    return result;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "policy-file", required_argument, NULL, 'o' },
+        { "session",     required_argument, NULL, 'S' },
+        { "context",     required_argument, NULL, 'c' },
+        { "auth-value",  required_argument, NULL, 'p' },
+    };
+
+    *opts = tpm2_options_new("o:S:c:p:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, 0);
+
+    return *opts != NULL;
+}
+
+bool is_input_option_args_valid(void) {
+
+    if (!ctx.session_path) {
+        LOG_ERR("Must specify -S session file.");
+        return false;
+    }
+
+    if (!ctx.flags.c) {
+        LOG_ERR("Must specify -c handle-id/ context file path.");
+        return false;
+    }
+
+    /*
+     *It is possible to have the authorization secret of the auth object NULL
+     */
```

Comment:
is this really going to be a common use case?

If you stall waiting on stdin, folks can use `ctrl + d` to indicate EOF to put in empty.

Reply:
They could also pass the empty string.

Also, you're argument handling should be arg count >= 0. If 0 args read stdin, else for each argument issue a policy password event.

Reply:
@williamcroberts added fmemopen to files handling routing to read from stdin. LMK. Thanks.